### PR TITLE
Ownership Inspection and Transference

### PR DESCRIPTION
**Pull request by KPrasch**
Based on PR #1146 

Introduces three new commands:

```
nucypher-deploy inspect --provider ...
```
```
nucypher-deploy transfer-ownership <contract name> --target-address ... --provider ... 
```
```
nucypher-deploy transfer-tokens <contract name> --target-address ... --provider ... 
```

### CODE REVIEW

File: nucypher/blockchain/eth/actors.py [hunk 1]
```diff
@@ -315,6 +320,31 @@ def deploy_network_contracts(self,
 
         return deployment_receipts
 
+    def relinquish_ownership(self,
```
Comment (szotov) [inline nucypher/blockchain/eth/actors.py]: Just idea: maybe extract "owner" functions to some base actor/agent

Comment (KPrasch) [reply nucypher/blockchain/eth/actors.py]: Yes, I *almost* wrote some kind of ownable agent mixin the other day.  Unsure if it can just be best represented as an attribute?

Comment (szotov) [reply nucypher/blockchain/eth/actors.py]: I think yes. Maybe same as upgradable attribute

File: nucypher/cli/actions.py [hunk 2]
```diff
@@ -100,7 +99,8 @@ def load_seednodes(emitter,
                    network_domains: set,
                    network_middleware: RestMiddleware = None,
                    teacher_uris: list = None
-                   ) -> List[Ursula]:
+                   ) -> list:
+    from nucypher.characters.lawful import Ursula
```
Comment (fjarri) [inline nucypher/cli/actions.py]: Was there a circular dependency?

File: nucypher/cli/deploy.py [hunk 3]
```diff
@@ -28,8 +28,8 @@
 from nucypher.cli import actions
 from nucypher.cli.actions import get_nucypher_password
 from nucypher.cli.actions import select_client_account
+from nucypher.cli.painting import paint_contract_deployment, paint_deployer_contract_status
```
Comment (fjarri) [inline nucypher/cli/deploy.py]: Sorry for nitpicking, but why not just add these imports to the list below?

File: nucypher/blockchain/eth/deployers.py [hunk 4]
```diff
@@ -382,10 +382,34 @@ def rollback(self, existing_secret_plaintext: bytes, new_secret_hash: bytes, gas
         return rollback_receipt
 
     def make_agent(self) -> EthereumContractAgent:
+        # TODO: Why is this commented out?
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/blockchain/eth/deployers.py]: :heavy_check_mark:

File: nucypher/cli/deploy.py [hunk 5]
```diff
@@ -246,20 +250,45 @@ def deploy(action,
                                                 allocation_outfile=allocation_outfile)
         return  # Exit
 
-    elif action == "transfer":
+    elif action == "transfer-tokens":
         token_agent = NucypherTokenAgent(blockchain=blockchain)
         missing_options = list()
-        if recipient_address is None:
+        if target_address is None:
             missing_options.append("--recipient-address")
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/cli/deploy.py]: :heavy_check_mark:

File: nucypher/cli/deploy.py [hunk 6]
```diff
@@ -246,20 +250,45 @@ def deploy(action,
...
-        if amount is None:
+        if value is None:
             missing_options.append("--amount")
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/cli/deploy.py]: :heavy_check_mark:

File: tests/cli/test_deploy.py [hunk 7]
```diff
@@ -13,11 +13,12 @@
     PolicyManagerAgent,
     AdjudicatorAgent,
     Agency,
-    EthereumContractAgent)
+    EthereumContractAgent
+)
+from nucypher.blockchain.eth.constants import STAKING_ESCROW_CONTRACT_NAME
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply tests/cli/test_deploy.py]: :heavy_check_mark:

File: nucypher/blockchain/eth/deployers.py [hunk 8]
```diff
@@ -382,10 +382,34 @@ def rollback(self, existing_secret_plaintext: bytes, new_secret_hash: bytes, gas
...
         #self.__check_policy_manager()  # Ensure the PolicyManager contract has already been initialized
         agent = self.agency(blockchain=self.blockchain, contract=self._contract)
         return agent
 
+    def transfer_ownership(self, new_owner: str, transaction_gas_limit: int = None):
```
[the remark this answers is not in the document]
Comment (derekpierre) [reply nucypher/blockchain/eth/deployers.py]: ^ @KPrasch any thoughts here?

[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/blockchain/eth/deployers.py]: Still thinking about this concept.  @szotov mentioned the same idea above.

File: nucypher/cli/deploy.py [hunk 9]
```diff
@@ -271,25 +275,24 @@ def deploy(action,
...
         if target_address is None:
-            missing_options.append("--recipient-address")
+            missing_options.append("--target-address")
         if value is None:
-            missing_options.append("--amount")
+            missing_options.append("--value")
         if missing_options:
             raise click.BadArgumentUsage(message=f"Need {' and '.join(missing_options)} to transfer tokens.")
 
         click.confirm(f"Transfer {value} from {token_agent.contract_address} to {target_address}?", abort=True)
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/cli/deploy.py]: Yeah. Well, in any case - how can I withdraw tokens from a contract?

File: nucypher/cli/deploy.py (hunk 3 above)
Comment (KPrasch) [reply nucypher/cli/deploy.py]: :heavy_check_mark:

File: nucypher/cli/deploy.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/cli/deploy.py]: :heavy_check_mark:

File: nucypher/blockchain/eth/interfaces.py [hunk 10]
```diff
@@ -350,9 +351,10 @@ def send_transaction(self,
         payload.update({'chainId': int(self.client.net_version),
                         'nonce': nonce,
                         'from': sender_address,
-                        'gasPrice': self.client.gas_price,
-                        # 'gas': 0,  # TODO: Gas Management
-                        })
+                        'gasPrice': self.client.gas_price})
+
+        if transaction_gas_limit:
+            payload['gas'] = int(transaction_gas_limit)
```
Comment (michwill) [inline nucypher/blockchain/eth/interfaces.py]: Good!

File: nucypher/blockchain/eth/agents.py [hunk 11]
```diff
@@ -111,7 +111,7 @@ def __repr__(self):
         return r.format(class_name, self.registry, self.registry_contract_name)
 
     def __eq__(self, other):
-        return bool(self.contract_address == other.contract_address)
+        return bool(self.contract.address == other.contract.address)
```
Comment (michwill) [inline nucypher/blockchain/eth/agents.py]: Why type casting? Is comparison ever not bool? Or is it a type hint?

**Comment by KPrasch** [comment]
Included in #1259

File: tests/blockchain/eth/entities/deployers/test_user_escrow_deployer.py [hunk 12]
```diff
@@ -22,61 +22,30 @@
...
-    assert deployment_progress.num_steps == len(user_escrow_proxy_deployer.deployment_steps) == 2
-    assert len(user_escrow_proxy_receipts) == 2
-
-    for step in user_escrow_proxy_deployer.deployment_steps:
-        assert user_escrow_proxy_receipts[step]['status'] == 1
-
-    deployer = UserEscrowDeployer(deployer_address=deployer_account,
-                                  registry=test_registry)
 
+    deployer = UserEscrowDeployer(deployer_address=testerchain.etherbase_account, registry=test_registry)
```
Comment (cygnusv) [inline tests/blockchain/eth/entities/deployers/test_user_escrow_deployer.py]: This changeset is getting rid of the original intention of this test, which is to check in detail a manual deployment of `UserEscrowProxy`. The test here should be analogous to what's in  `test_policy_manager_deployer`, which using first a `staking_escrow_deployer` fixture, tests a manual deployment of `PolicyManager`. Translating to `UserEscrowProxy` means that the test must use a `staking_escrow_deployer` and `policy_manager_deployer` fixture in order to test manually a deployment of `UserEscrowProxy`.

In other words, IMO these tests shouldn't use the `agency` fixture.

File: tests/blockchain/eth/entities/deployers/test_user_escrow_deployer.py [hunk 13]
```diff
@@ -22,61 +22,30 @@
...
     deployer_account = testerchain.etherbase_account
 
-    linker_deployer = LibraryLinkerDeployer(registry=test_registry,
-                                            deployer_address=deployer_account,
-                                            target_contract=user_escrow_proxy_deployer.contract,
-                                            bare=True)
-    linker_address = linker_deployer.contract_address
-
+    linker = testerchain.get_contract_by_name(registry=test_registry, name=LibraryLinkerDeployer.contract_name)
+    linker_address = linker.address
```
Comment (cygnusv) [inline tests/blockchain/eth/entities/deployers/test_user_escrow_deployer.py]: 👍 much cleaner

File: nucypher/blockchain/economics.py [hunk 14]
```diff
@@ -195,3 +195,6 @@ def deployment_parameters(self) -> Tuple[int, ...]:
         ]
 
         return tuple(map(int, deployment_parameters))
+
+
+Economics = Union[SlashingEconomics, TokenEconomics]
```
Comment (cygnusv) [inline nucypher/blockchain/economics.py]: Classy! although it may go away after the rebase...

File: nucypher/blockchain/eth/actors.py [hunk 15]
```diff
@@ -142,14 +142,19 @@ class ContractAdministrator(NucypherTokenActor):
         NucypherTokenDeployer,
     )
 
-    upgradeable_deployer_classes = (
+    dispatched_upgradeable_deployer_classes = (
```
Comment (cygnusv) [inline nucypher/blockchain/eth/actors.py]: I find it clearer a name like `deployer_classes_with_dispatcher`.

File: tests/blockchain/eth/entities/deployers/test_user_escrow_deployer.py (hunk 12 above)
Comment (KPrasch) [reply tests/blockchain/eth/entities/deployers/test_user_escrow_deployer.py]: Yeah - I Agree.  I must have overlooked this.

File: nucypher/blockchain/economics.py (hunk 14 above)
Comment (KPrasch) [reply nucypher/blockchain/economics.py]: :heavy_check_mark:

File: nucypher/blockchain/eth/agents.py (hunk 11 above)
Comment (KPrasch) [reply nucypher/blockchain/eth/agents.py]: Just a style habit  I have for `__eq__`.

File: tests/blockchain/eth/entities/deployers/test_user_escrow_deployer.py (hunk 12 above)
Comment (cygnusv) [reply tests/blockchain/eth/entities/deployers/test_user_escrow_deployer.py]: It's looking better, but still needs to check the details of the `UserEscrowDeployer`(which was the point of the original test). In other words, these lines are missing:
```python
user_escrow_proxy_receipts = user_escrow_proxy_deployer.deploy(secret_hash=secret_hash,
                                                                   progress=deployment_progress)
# deployment steps must match expected number of steps
assert deployment_progress.num_steps == len(user_escrow_proxy_deployer.deployment_steps) == 2
assert len(user_escrow_proxy_receipts) == 2

for step in user_escrow_proxy_deployer.deployment_steps:
        assert user_escrow_proxy_receipts[step]['status'] == 1
```

Comment (KPrasch) [reply tests/blockchain/eth/entities/deployers/test_user_escrow_deployer.py]: :+1:

Comment (KPrasch) [reply tests/blockchain/eth/entities/deployers/test_user_escrow_deployer.py]: :heavy_check_mark:

File: tests/blockchain/eth/entities/deployers/test_user_escrow_deployer.py [hunk 16]
```diff
@@ -18,38 +18,53 @@
...
-@pytest.fixture(scope='module')
-def user_escrow_proxy_deployer(testerchain, session_agency, test_registry):
-    print("ENTER USER ESCROW")
-    testerchain = testerchain
-    deployer = testerchain.etherbase_account
-    user_escrow_proxy_deployer = UserEscrowProxyDeployer(deployer_address=deployer,
-                                                         registry=test_registry)
-    return user_escrow_proxy_deployer
+@pytest.mark.slow()
+def test_user_escrow_proxy_deployer(testerchain, deployment_progress, test_registry):
```
Comment (cygnusv) [inline tests/blockchain/eth/entities/deployers/test_user_escrow_deployer.py]: good change 👍

File: tests/cli/test_deploy.py [hunk 17]
```diff
@@ -34,11 +37,12 @@ def generate_insecure_secret() -> str:
 
 
 @pytest.fixture(scope="module")
-def registry_filepath():
+def registry_filepath(test_registry):
     # TODO: Use temp module
     registry_filepath = os.path.join('tmp', 'nucypher-deploy-test.json')
     if os.path.exists(registry_filepath):
         os.remove(registry_filepath)
+    # test_registry.commit(registry_filepath)
```
Comment (cygnusv) [inline tests/cli/test_deploy.py]: get rid of this line?

File: tests/fixtures.py [hunk 18]
```diff
@@ -418,20 +412,23 @@ def _make_agency(testerchain, test_registry):
...
+
+    user_escrow_proxy_deployer = UserEscrowProxyDeployer(deployer_address=origin, registry=test_registry)
+    user_escrow_proxy_deployer.deploy(secret_hash=INSECURE_DEPLOYMENT_SECRET_HASH)
 
-    token_agent = token_deployer.make_agent()              # 1 Token
-    staking_agent = staking_escrow_deployer.make_agent()   # 2 Staking Escrow
-    policy_agent = policy_manager_deployer.make_agent()    # 3 Policy Agent
-    _adjudicator_agent = adjudicator_deployer.make_agent()  # 4 Adjudicator
+    token_agent = token_deployer.make_agent()                           # 1 Token
+    staking_agent = staking_escrow_deployer.make_agent()                # 2 Miner Escrow
```
Comment (cygnusv) [inline tests/fixtures.py]:
```suggestion
    staking_agent = staking_escrow_deployer.make_agent()                # 2 StakingEscrow
```

Comment (KPrasch) [reply tests/fixtures.py]: :question:

File: nucypher/blockchain/eth/deployers.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/blockchain/eth/deployers.py]: See #1314

File: nucypher/blockchain/eth/actors.py (hunk 1 above)
Comment (KPrasch) [reply nucypher/blockchain/eth/actors.py]: See #1314

**Comment by KPrasch** [comment]
Follow-up PR - See #1314

File: tests/fixtures.py (hunk 18 above)
Comment (cygnusv) [reply tests/fixtures.py]: Go to the end of the line. There's a "Miner Escrow"

File: nucypher/cli/painting.py [hunk 19]
```diff
@@ -173,36 +178,138 @@ def paint_known_nodes(emitter, ursula) -> None:
...
+
+Provider URI ............. {blockchain.provider_uri}
+Registry  ................ {administrator.registry.filepath}
+
+* Standard Deployments
+=====================================================================
+
+NucypherToken ........... {token_agent.contract_address}
+    ~ Ethers ............ {blockchain.client.get_balance(token_agent.contract_address)} wei
+    ~ Tokens ............ {token_agent.get_balance(token_agent.contract_address)} wei"""
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/cli/painting.py]: :heavy_check_mark:

File: nucypher/cli/painting.py [hunk 20]
```diff
@@ -173,36 +178,138 @@ def paint_known_nodes(emitter, ursula) -> None:
...
+        user_escrow_payload = f"""
+UserEscrowProxy .......... {bare_contract.address}
+    ~ LibraryLinker ...... {linker_deployer.contract.address}
+        ~ Owner .......... {linker_deployer.contract.functions.owner().call()}
+        ~ Target ......... {linker_deployer.contract.functions.target().call()}"""
+        emitter.echo(user_escrow_payload)
+        emitter.echo(sep)
+
+    except BaseContractRegistry.UnknownContract:
+        message = f"UserEscrowProxy is not enrolled in {administrator.registry.filepath}"
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/cli/painting.py]: :heavy_check_mark:

File: nucypher/cli/deploy.py [hunk 21]
```diff
@@ -264,23 +273,47 @@ def deploy(action,
...
+                raise click.Abort()
+            else:
+                contract_deployer = contract_deployer_class(registry=ADMINISTRATOR.registry,
+                                                            deployer_address=ADMINISTRATOR.deployer_address)
+                receipt = contract_deployer.transfer_ownership(new_owner=target_address, transaction_gas_limit=gas)
+                emitter.ipc(receipt, request_id=0, duration=0)  # TODO: eh?
+                return  # Exit
+        else:
+            receipts = ADMINISTRATOR.relinquish_ownership(new_owner=target_address, transaction_gas_limit=gas)
+            emitter.ipc(receipts, request_id=0, duration=0)  # TODO: eh?
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/cli/deploy.py]: :heavy_check_mark:

File: nucypher/cli/deploy.py [hunk 22]
```diff
@@ -264,23 +273,47 @@ def deploy(action,
...
         token_agent = ContractAgency.get_agent(NucypherTokenAgent, registry=registry)
         missing_options = list()
-        if recipient_address is None:
-            missing_options.append("--recipient-address")
-        if amount is None:
-            missing_options.append("--amount")
+        if target_address is None:
+            missing_options.append("--target-address")
+        if value is None:
+            missing_options.append("--value")
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/cli/deploy.py]: :heavy_check_mark:

File: nucypher/cli/painting.py [hunk 23]
```diff
@@ -173,36 +178,138 @@ def paint_known_nodes(emitter, ursula) -> None:
...
+    emitter.echo(sep)
+    emitter.echo(contracts)
+    emitter.echo(sep)
+    emitter.echo(blockchain)
+    emitter.echo(sep)
+    emitter.echo(staking)
+    emitter.echo(sep)
+
+
+def paint_deployer_contract_status(emitter, administrator) -> None:
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/cli/painting.py]: :heavy_check_mark:

File: nucypher/blockchain/eth/deployers.py [hunk 24]
```diff
@@ -654,6 +701,13 @@ def deploy(self, secret_hash: bytes, gas_limit: int = None, progress=None) -> di
         self._contract = user_escrow_proxy_contract
         return deployment_receipts
 
+    @classmethod
+    def get_latest_version(cls, blockchain) -> Contract:
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/blockchain/eth/deployers.py]: See #1314

File: tests/cli/test_deploy.py (hunk 17 above)
Comment (KPrasch) [reply tests/cli/test_deploy.py]: :heavy_check_mark:

File: nucypher/blockchain/eth/actors.py (hunk 15 above)
Comment (KPrasch) [reply nucypher/blockchain/eth/actors.py]: Will follow up in #1314

File: tests/fixtures.py (hunk 18 above)
Comment (KPrasch) [reply tests/fixtures.py]: :heavy_check_mark:
